File: emu/emu_plot/python/plot_msim.py
# Read Modified Simulation Tool output

import os
import numpy as np
import glob
import matplotlib.pyplot as plt
import global_emu_var as emu
import lib_python as emupy

# Create a custom colormap that maps masked values to gray
cmap = plt.cm.jet
cmap.set_bad(color='gray')  # Set the color for masked elements to gray

# --------------------------------------------
def plot_msim(frun):
    """
    Check Modified Simulation Tool output.
    
    Parameters:
    frun (str): The main run directory path.
    """
    
    # Set EMU output file directory
    frun_output = os.path.join(frun, 'diags')
    
    # Search files
    print('')
    print("Checking EMU standard model state output ... ")
    print('')
    
    # Assuming plot_state is a function that processes the data and populates the naa variables
    naa_2d_day, naa_2d_mon, naa_3d_mon = emupy.plot_state(frun_output)
    
    ndum = max(naa_2d_day, naa_2d_mon, naa_3d_mon)
    print('')
    print('*********************************************')
    
    if ndum != 0:
        print("EMU's standard model state output can be sampled using EMU's ")
        print("Sampling Tool, specifying the diag subdirectory of this run")
        print("in response to the Tool's prompt;")
        print(frun_output)
        print('')
    else:
        print("No diagnostic state output found in this run's diag subdirectory.")
        print(frun_output)
        print('')
    
    # Search subdirectories
    print('')
    print('*********************************************')
    print("Checking Budget output ... ")
    
    # Search for all subdirectories
    entries = glob.glob(os.path.join(frun_output, '*'))
    entries.sort()
    
    # Initialize the counter
    subdir_count = 0
    
    # Loop through the entries and count only directories
    for entry in entries:
        if os.path.isdir(entry):
            subdir_count += 1
    
    if subdir_count != 0:
        # Count total number of .data files in the subdirectories
        files = glob.glob(os.path.join(frun_output, '*/*.data'))
        files.sort()
        total_file_count = len(files)
        
        if total_file_count != 0:
            # Output # of subdirectories
            print('')
            print(f'Total number of subdirectories: {subdir_count}')
            
            # Initialize the second counter
            subdir_count2 = 0
            print('')
            
            # Loop through each subdirectory again and count number of .data files
            for entry in entries:
                if os.path.isdir(entry):
                    subdir_count2 += 1
                    
                    # Search for all files ending with .data in the directory
                    files = glob.glob(os.path.join(entry, '*.data'))
                    files.sort()
                    
                    # Count the number of .data files
                    file_count = len(files)
                    
                    # Print name of subdirectory and number of .data files
                    fname = os.path.basename(entry)
                    fdum = f'   {subdir_count2}) {fname} has {file_count} files'
                    print(fdum)
            
            print('')
            print('*********************************************')
            print("Budget output of this run can be analyzed using ")
            print("EMU's Budget Tool, specifying the diag subdirectory of this run")
            print("in response to the Tool's prompt;")
            print(os.path.join(frun, 'diags'))
            print('')
        else:
            print('')
            print("No budget output found in this run's diag subdirectory.")
            print(os.path.join(frun, 'diags'))
            print('')
    else:
        print("No budget output found in this run's diag subdirectory.")
        print(os.path.join(frun, 'diags'))
        print('')

File: emu/emu_plot/matlab/plot_msim.m
function plot_msim(frun)
% Check Modified Simulation Tool output

% ---------------
% Set EMU output file directory
frun_output = fullfile(frun, 'diags');

% ---------------
% Search files

fprintf('\nChecking EMU standard model state output ... \n\n');

% Call plot_state function
[naa_2d_day, naa_2d_mon, naa_3d_mon] = plot_state(frun_output);

ndum = max([naa_2d_day, naa_2d_mon, naa_3d_mon]);

fprintf('\n*********************************************\n');
if ndum ~= 0
    fprintf('EMU''s standard model state output can be sampled using EMU''s\n');
    fprintf('Sampling Tool, specifying the diag subdirectory of this run\n');
    fprintf('in response to the Tool''s prompt;\n');
    fprintf('%s\n\n', frun_output);
else
    fprintf('No diagnostic state output found in this run''s diag subdirectory.\n');
    fprintf('%s\n\n', frun_output);
end

% ---------------
% Search subdirectories

fprintf('\n*********************************************\n');
fprintf('Checking Budget output ...\n');

% List entries in diag directory
entries = dir(frun_output);
entries = entries([entries.isdir]);  % Keep only directories
entries = entries(~ismember({entries.name}, {'.', '..'}));  % Exclude . and ..

subdir_count = numel(entries);

if subdir_count ~= 0
    % Count total number of .data files in subdirectories
    data_files = dir(fullfile(frun_output, '*', '*.data'));
    total_file_count = numel(data_files);

    if total_file_count ~= 0
        fprintf('\nTotal number of subdirectories: %d\n\n', subdir_count);

        subdir_count2 = 0;
        for i = 1:subdir_count
            subdir_path = fullfile(frun_output, entries(i).name);
            files = dir(fullfile(subdir_path, '*.data'));
            file_count = numel(files);

            subdir_count2 = subdir_count2 + 1;
            fprintf('   %d) %s has %d files\n', ...
                subdir_count2, entries(i).name, file_count);
        end

        fprintf('\n*********************************************\n');
        fprintf('Budget output of this run can be analyzed using\n');
        fprintf('EMU''s Budget Tool, specifying the diag subdirectory of this run\n');
        fprintf('in response to the Tool''s prompt;\n');
        fprintf('%s\n\n', fullfile(frun, 'diags'));
    else
        fprintf('\nNo budget output found in this run''s diag subdirectory.\n');
        fprintf('%s\n\n', fullfile(frun, 'diags'));
    end
else
    fprintf('No budget output found in this run''s diag subdirectory.\n');
    fprintf('%s\n\n', fullfile(frun, 'diags'));
end
end
